### PR TITLE
 Resize cephfs CSI volumes

### PR DESCRIPTION
**Pull request by humblec**
Resize CephFS Volumes
    
    This feature enables CephFS Volume expansion on demand
    based on the CO resizer request.

    Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by Madhu-1** [comment]
@humblec can you some more description to this PR, what issue we are going to solve?

**Comment by Madhu-1** [comment]
adding DNM  label. till it complete

**Comment by humblec** [comment]
@Madhu-1 I have updated the commit message.

**Comment by humblec** [comment]
@dillaman Have a look at this PR, this is in better shape :)

File: pkg/cephfs/util.go [hunk 1]
```diff
@@ -126,6 +126,24 @@ func (cs *ControllerServer) validateCreateVolumeRequest(req *csi.CreateVolumeReq
 	return nil
 }
 
+// Controller expand volume request validation
+func (cs *ControllerServer) validateExpandVolumeRequest(req *csi.ControllerExpandVolumeRequest) error {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_EXPAND_VOLUME); err != nil {
+		return fmt.Errorf("invalid ExpandVolumeRequest: %v", err)
```
Comment (ShyamsundarR) [inline pkg/cephfs/util.go]: Return a status.Error instead of a raw error string?

File: pkg/cephfs/volume.go [hunk 2]
```diff
@@ -63,6 +78,19 @@ func createVolume(volOptions *volumeOptions, adminCr *credentials, volID volumeI
 
 	if pathExists(volRoot) {
 		klog.V(4).Infof("cephfs: volume %s already exists, skipping creation", volID)
+		size, err := getVolumeAttribute(volRoot, "ceph.quota.max_bytes")
```
Comment (ShyamsundarR) [inline pkg/cephfs/volume.go]: If path does not exist, the call from ControllerExpandVolume would go ahead and attempt to create the path and perform other operations as needed. It is better to deal with an expand request as a separate call, and not fold the operation into createVolume, keeping abstractions clean. The code reuse (by using createVolume) seems minimal and hence not really needed.

File: pkg/cephfs/controllerserver.go [hunk 3]
```diff
@@ -44,6 +44,34 @@ var (
...
 
+// ExpandVolume expand CephFS Volumes on demand based on resizer request
+func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
+
+	if err := cs.validateExpandVolumeRequest(req); err != nil {
+		klog.Errorf("ControllerExpandVolumeRequest validation failed: %v", err)
+		return nil, err
+	}
+
+	volID := req.GetVolumeId()
```
Comment (ShyamsundarR) [inline pkg/cephfs/controllerserver.go]: Should we allow volume expansion for volumes that CSI did not provision? IOW, when a volume create came in with the provision as false?

File: pkg/cephfs/controllerserver.go [hunk 4]
```diff
@@ -44,6 +44,34 @@ var (
...
+	secret := req.GetSecrets()
```
Comment (ShyamsundarR) [inline pkg/cephfs/controllerserver.go]: The resizer secret [name and namespace](https://github.com/kubernetes-csi/external-resizer/blob/c398721342811a9b47f8b96ff7112288e57d465b/pkg/resizer/csi_resizer.go#L40) are specialized, which in turn would need the same in the storageclass. Request update of documentation and examples to reflect the same.

**Comment by ajarr** [comment]
@humblec @Madhu-1 close this in favor of https://github.com/ceph/ceph-csi/pull/412?

**Comment by Madhu-1** [comment]
closing this in favor of  #412, @humblec if require please feel free to reopen

**Pull request closed without merging**